extensions/wikidata/src/org/openrefine/wikidata/editing/ConnectionManager.java

```diff
@@ -0,0 +1,148 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ ******************************************************************************/
+package org.openrefine.wikidata.editing;
+
+import java.io.IOException;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.wikidata.wdtk.wikibaseapi.ApiConnection;
+import org.wikidata.wdtk.wikibaseapi.LoginFailedException;
+
+import com.google.refine.ProjectManager;
+import com.google.refine.preference.PreferenceStore;
+
+/**
+ * Manages a connection to Wikidata, with login credentials stored in the
+ * preferences.
+ * 
+ * Ideally, we should store only the cookies and not the password. But
+ * Wikidata-Toolkit does not allow for that as cookies are kept private.
+ * 
+ * This class is also hard-coded for Wikidata: generalization to other Wikibase
+ * instances should be feasible though.
+ * 
+ * @author Antonin Delpeuch
+ */
+
+public class ConnectionManager {
+
+    public static final String PREFERENCE_STORE_KEY = "wikidata_credentials";
+
+    private PreferenceStore prefStore;
+    private ApiConnection connection;
+
+    private static class ConnectionManagerHolder {
+
+        private static final ConnectionManager instance = new ConnectionManager();
+    }
+
+    public static ConnectionManager getInstance() {
+        return ConnectionManagerHolder.instance;
+    }
+
+    private ConnectionManager() {
+        prefStore = ProjectManager.singleton.getPreferenceStore();
+        connection = null;
+        restoreSavedConnection();
+    }
+
+    public void login(String username, String password, boolean rememberCredentials) {
+        if (rememberCredentials) {
+            try {
+                JSONArray array = new JSONArray();
+                JSONObject obj = new JSONObject();
+                obj.put("username", username);
+                obj.put("password", password);
+                array.put(obj);
+                prefStore.put(PREFERENCE_STORE_KEY, array);
+                // TODO save preferences (find out how)
```

Comment:
No saved? Maybe a cookie is an option?

Reply:
This comment is actually out of date - OR saves preferences itself. Thanks for picking that up though.